runtime/src/bank/partitioned_epoch_rewards/calculation.rs

```diff
@@ -375,71 +383,135 @@ impl Bank {
+                        )
+                    },
+                    |(mut vote_rewards, len_stake_rewards_some, total_stake_rewards),
+                     (vote_pubkey, vote_reward, stakers_reward)| {
+                        vote_rewards
+                            .entry(vote_pubkey)
+                            .and_modify(|dst_vote_reward: &mut VoteReward| {
+                                dst_vote_reward.vote_rewards = dst_vote_reward
+                                    .vote_rewards
+                                    .saturating_add(vote_reward.vote_rewards)
+                            })
+                            .or_insert(vote_reward);
+                        (
+                            vote_rewards,
+                            len_stake_rewards_some.saturating_add(1),
+                            total_stake_rewards.saturating_add(stakers_reward),
+                        )
+                    },
+                )
+                .reduce(
+                    || {
+                        (
+                            HashMap::with_hasher(PubkeyHasherBuilder::default()),
+                            usize::default(),
+                            u64::default(),
+                        )
+                    },
+                    |(vote_rewards_a, len_stake_rewards_some_a, total_stake_rewards_a),
+                     (vote_rewards_b, len_stake_rewards_some_b, total_stake_rewards_b)| {
+                        let (mut dst_vote_rewards, src_vote_rewards) =
+                            if vote_rewards_a.len() >= vote_rewards_b.len() {
+                                (vote_rewards_a, vote_rewards_b)
+                            } else {
+                                (vote_rewards_b, vote_rewards_a)
+                            };
+                        for (vote_pubkey, vote_reward) in src_vote_rewards {
+                            dst_vote_rewards
+                                .entry(vote_pubkey)
+                                .and_modify(|dst_vote_reward: &mut VoteReward| {
+                                    dst_vote_reward.vote_rewards = dst_vote_reward
+                                        .vote_rewards
+                                        .saturating_add(vote_reward.vote_rewards)
+                                })
+                                .or_insert(vote_reward);
+                        }
+                        (
+                            dst_vote_rewards,
+                            len_stake_rewards_some_a.saturating_add(len_stake_rewards_some_b),
+                            total_stake_rewards_a.saturating_add(total_stake_rewards_b),
+                        )
+                    },
+                )
+        });
+        // SAFETY: We initialized all the `stake_rewards` elements up to the capacity.
+        unsafe {
+            stake_rewards.set_len(stake_rewards.capacity());
+            stake_rewards.set_len_some(len_stake_rewards_some);
```

Comment:
This addresses https://github.com/anza-xyz/agave/pull/6900#pullrequestreview-3158696020